kafka-connect-s3/docs/s3_connector.rst

```diff
@@ -251,6 +253,83 @@ Configuration
 This section gives example configurations that cover common scenarios. For detailed description of all the
 available configuration options of the S3 connector go to :ref:`Configuration Options<s3_configuration_options>`
 
+Configurable credential provider
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Some use cases require more fine-grained credentials configuration for AWS so that each connector
+could have its own credentials.
+
+To use configurable credential provider, ``s3.credentials.provider.class`` should be a class which
+implements both ``com.amazonaws.auth.AWSCredentialsProvider`` and ``org.apache.kafka.common
+.Configurable`` interfaces.
+
+A subset of properties, starting from ``s3.credentials.provider.`` (with stripped out prefix)
+will be passed to an instance of a credential provider.
+
+Here is a non-thread safe reference implementation of configurable credential provider:
+
+.. sourcecode:: java
+
+    package com.example;
+
+    import com.amazonaws.auth.AWSCredentials;
+    import com.amazonaws.auth.AWSCredentialsProvider;
+    import com.amazonaws.auth.BasicAWSCredentials;
+    import org.apache.kafka.common.Configurable;
+    import org.apache.kafka.common.config.ConfigException;
+    import java.util.Map;
+
+    public class ConfigurableCredentialsProvider implements AWSCredentialsProvider, Configurable {
+
+      private static final String ACCESS_KEY_NAME = "key";
+      private static final String SECRET_KEY_NAME = "secret";
+      private AWSCredentials credentials;
+      private boolean configured = false;
+
+      @Override
+      public AWSCredentials getCredentials() {
+        if (configured) {
+          return credentials;
+        } else {
+          throw new IllegalStateException("Credential provider is not configured");
+        }
+      }
+
+      @Override
+      public void refresh() {}
+
+      @Override
+      public void configure(final Map<String, ?> configs) {
+
+        final String accessKeyId = (String) configs.get(ACCESS_KEY_NAME);
+        final String secretKey = (String) configs.get(SECRET_KEY_NAME);
+
+        validateConfigs(configs);
+
+        credentials = new BasicAWSCredentials(accessKeyId, secretKey);
+        configured = true;
+      }
+
+      private void validateConfigs(Map<String, ?> configs) {
+
+        if (!configs.containsKey(ACCESS_KEY_NAME)
+            || !configs.containsKey(SECRET_KEY_NAME)) {
+          throw new ConfigException(String.format("%s and %s are mandatory configuration properties",
+              ACCESS_KEY_NAME, SECRET_KEY_NAME
+          ));
+        }
+      }
+    }
+
+with following configurations:
+
+.. sourcecode:: bash
+
+
+  s3.credentials.provider.class=com.example.ConfigurableCredentialsProvider
+  s3.credentials.provider.key=YOUR_ACCESS_KEY_ID
+  s3.credentials.provider.secret=YOUR_SECRET_ACCESS_KEY
```

Comment:
This example could go into the JavaDoc of the `SampleCredentialsProvider`.

Reply:
Do we have JavaDoc published somewhere? I have found this - https://docs.confluent.io/current/connect/javadocs/index.html.